sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/EventGridEvent.java

```diff
@@ -48,57 +68,88 @@
+     * @param eventGridJsonString the JSON string containing one or more EventGridEvent objects.
+     *
+     * @return A list of {@link EventGridEvent EventGridEvents} deserialized from eventGridJsonString.
      * @throws IllegalArgumentException if eventGridJsonString isn't a JSON string for a eventgrid event
      * or an array of it.
      * @throws NullPointerException if eventGridJsonString is {@code null}.
+     * @throws IllegalArgumentException if the {eventGridJsonString isn't a Json string or can't be deserialized
+     * into valid EventGridEvent instances.
      */
     public static List<EventGridEvent> fromString(String eventGridJsonString) {
         try {
-            return Arrays.stream(DESERIALIZER
-                .deserialize(new ByteArrayInputStream(eventGridJsonString.getBytes(StandardCharsets.UTF_8)),
-                    TypeReference.createInstance(com.azure.messaging.eventgrid.implementation.models.EventGridEvent[].class)))
+            return SERIALIZER.deserialize(
+                new ByteArrayInputStream(eventGridJsonString.getBytes(StandardCharsets.UTF_8)),
+                DESERIALIZER_TYPE_REFERENCE)
+                .stream()
                 .map(internalEvent -> {
                     if (internalEvent.getSubject() == null || internalEvent.getEventType() == null
-                        || internalEvent.getData() == null) {
-                        throw logger.logExceptionAsError(new IllegalArgumentException(
-                            "'subject', 'type', and 'data' are mandatory attributes for an EventGridEvent. " +
-                                "Check if the input param is a JSON string for an EventGridEvent or an array of it."));
+                        || internalEvent.getData() == null || internalEvent.getDataVersion() == null) {
+                        List<String> nullAttributes = new ArrayList<String>();
+                        if (internalEvent.getSubject() == null) {
+                            nullAttributes.add("'subject'");
+                        }
+                        if (internalEvent.getEventType() == null) {
+                            nullAttributes.add("'eventType'");
+                        }
+                        if (internalEvent.getData() == null) {
+                            nullAttributes.add("'data'");
+                        }
+                        if (internalEvent.getData() == null) {
+                            nullAttributes.add("'dataVersion'");
+                        }
+                        throw LOGGER.logExceptionAsError(new IllegalArgumentException(
+                            "'subject', 'eventType', 'data' and 'dataVersion' are mandatory attributes for an "
+                                + "EventGridEvent. This Json string has null values for "
+                                + String.join(",", nullAttributes)
+                                + ". Please make sure the input Json string has the required attributes"));
```

Comment:
This is a good error message to make it more diagnosable! Additionally, since the jsonstring contains a list of events, it may be worth including the index of the event that is doesn't have the required attribute. "The nth event in the json string has null values for .."

Reply:
Added it.